device_attributes/index.bs

```diff
@@ -0,0 +1,192 @@
+Repository: https://github.com/WICG/WebApiDevice
+Group: wicg
+</pre>
+ 
+<pre class=link-defaults>
+spec:html; type:interface; text:Navigator
+spec:url; type:dfn; text:origin
+spec:html; type:dfn; for:/; text:global object
+spec:webidl; type:dfn; text:resolve
+spec:webidl; type:dfn; text:record
+</pre>
+ 
+<pre class="anchors">
+ 
+</pre>
+ 
+# Model # {#model}
+ 
+## Managed devices ## {#managed-devices}
+ 
+The API is presumed to be used on devices which are not fully controlled by the end user, but rather by an external entity, the device administrator. Device administrators are given the power to fully control the managed devices.
+ 
+## Managed web applications ## {#managed-web-applications}
+ 
+The API is presumed to be used by web applications, which are understood and managed by the device administrators.
+ 
+## Device attributes ## {#device-attributes}
+ 
+The API is presumed to be used with a management infrastructure which allows administrators to manage a fleet of devices and configure properties for those devices.
+ 
+### <dfn>Annotated Asset ID</dfn> ### {#annotated-asset-id}
+
+An administrator-defined value which uniquely identifies a device within an organization.
+
+### <dfn>Annotated Location</dfn> ### {#annotated-location}
+
+An administrator-defined value which uniquely identifies a location within an organization.
+
+### <dfn>Directory ID</dfn> ### {#directory-id}
+ 
+An inventory management system-defined value which uniquely identifies a device within an organization.
+ 
+### <dfn>Hostname</dfn> ### {#hostname}
+ 
+An administrator-defined value which is used as the device hostname during DHCP requests.
+ 
+### <dfn>Serial Number</dfn> ### {#serial-number}
+ 
+A manufacturer-defined value which uniquely identifies a device among those produced by that manufacturer.
+ 
+# {{NavigatorManagedData}} interface # {#navigatormanageddata-interface}
+ 
+<xmp class="idl">
+[
+  SecureContext,
+  Exposed=Window
+] interface NavigatorManagedData : EventTarget {
+  // Device Attributes API.
+  Promise<DOMString> getAnnotatedAssetId();
+  Promise<DOMString> getAnnotatedLocation();
+  Promise<DOMString> getDirectoryId();
+  Promise<DOMString> getHostname();
+  Promise<DOMString> getSerialNumber();
+};
+</xmp>
+ 
+Methods on this interface typically complete asynchronously, queuing work on the <dfn>managed data task source</dfn>.
+ 
+## {{getAnnotatedAssetId()}} method ## {#getAnnotatedAssetId-method}
+ 
+  The <dfn method for="NavigatorManagedData">getAnnotatedAssetId()</dfn> method steps are:
+  1. Let |promise| be [=a new promise=].
+  1. Run the following steps in parallel:
+      1. If this API is not called by a managed web application with permission to access device attributes, reject |promise| with a {{NotAllowedError}} DOMException.
+      1. If no [=Annotated Asset ID=] has been set by the administrator, resolve |promise| with {{undefined}} value.
```

Comment:
In the case of a missing asset ID, you resolve the promise in parallel and then also queue a task to resolve it with the missing value.

Reply:
More specifically, you need to:

1. Add the phrase "and abort these steps" to the end of any action which should cause the algorithm to finish early because of an error.
2. Always queue a task when resolving or rejecting a Promise.

These are both things I missed when reviewing the steps for getManagedConfiguration().

Reply:
Rephrase the steps as suggested. Please note that missing asset ID is not treated as an exception. The promise will still be solved with an 'undefined' value.